[enh] add pyrightconfig.json and integrate pyright into CI

not-my-profile opened the pull request:
By adding this file Pyright automatically detects the packages from our Python virtual environment.

This can be tested by using the Pyright extension for VS Code or by running `npx pyright`.

not-my-profile committed:
[enh] add pyrightconfig.json

By adding this file Pyright automatically detects the
packages from our Python virtual environment.

This can be tested by using the Pyright extension
for VS Code or by running npx pyright.

return42 committed:
[fix] pyrightconfig.json include only dedicated folders in the test

Without specifying folders to check, pyright will also scan files in folders
like ./build, ./cache, ./.nvm and more.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] add pyright to nvm's node installation

Pyright [1] is in the nvm enviroment, may be you need to rebuild the nvm
environment and install nodejs in::

    ./manage nvm.clean
    ./manage nvm.nodejs

The last command installs nodejs and the packages from .nvm_packages.

You can test your pyright installation, to get a bash within the nvm run::

   ./manage nvm.bash

   (nvm) $ which pyright
   ./.nvm/versions/node/v16.13.0/bin/pyright

If you have a local nvm in your HOME folder, the output from ``which`` is
different.  Press ``[CTRL-D]`` to get out of this bash.

[1] https://github.com/microsoft/pyright

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] manage test.pyright: a commandline to run pyright tests

This patch implements the command and Makefile target::

    ./manage test.pyright
    make test.pyright

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[emacs] .dir-locals.el: activate pyright in python-mode

To get in use of pyright type cheker in Emacs, a pyright installation [1] is
needed and in Emacs the lsp-pyright package has to be installed::

    M-x package-install lsp-pyright

[1] https://github.com/Microsoft/pyright
[2] https://github.com/emacs-lsp/lsp-pyright

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

not-my-profile committed:
[enh] clarify why we run pyright in the pyenv

return42 commented on manage:
```diff
@@ -687,7 +687,7 @@ test.pylint() {
     build_msg TEST "[pyright] static type check of python sources"
     nodejs.ensure
-    pyenv.cmd pyright
+    pyright
```
Why do you think `pyenv.cmd` is not needed here?

not-my-profile replied:
Why do you think it is needed here? :P Pyright isn't written in Python, it's written in Typescript.

return42 replied:
> Pyright isn't written in Python, it's written in Typescript.

This was what I needed to know, to give the right feedback: the python environment (virtualenv) is needed ...

```
{
    "venvPath": "local",
    "venv": "py3",
...
```

... even if it is a static type checker does not need the python interpreter itself.

with the pyenv.cmd we ensure that the virtualenv in `./local` is build.

By example first clean an than run that test target ..

    make clean test.pyright

In such cases there is no `./local/py3` and pyright alternatively uses the python environment from your HOME or if there is no special environment in your HOME it uses the python environment from the OS.

not-my-profile replied:
Ah yeah that makes sense. I still think that it would be clearer to just do a `pyenv.install` before.

not-my-profile replied:
I rebased the PR accordingly.

return42 replied:
> Ah yeah that makes sense. I still think that it would be clearer to just do a `pyenv.install` before.

Sadly this is not enough .. `pyenv.install`  will install virtualenv but not activate.  Don't mix python environment with the nvm environment where no activation is needed.

not-my-profile replied:
The virtualenv doesn't need to be active for pyright ... it's a static code analyzer written in TypeScript, it does not actually execute any Python code.

return42 replied:
> The virtualenv doesn't need to be active for pyright ... it's a static code analyzer written in TypeScript, it does not actually execute any Python code.

Even if it is static, the analyser needs to import the modules from the environment to get the types.

not-my-profile replied:
Yes but afaik Pyright discovers the modules itself via the `venvPath` & `venv` settings in `pyrightconfig.json`. Afaik it does not use (or need) environment variables to do so.

return42 replied:
> Afaik it does not use (or need) environment variables to do so.

"AFAIK" .. this is what I also know/thought .. but we don't know it .. and I think for those do not know the details, its hard to understand if we use the python environment for all py-processes but not for the analyzer. The cost is not high so I think it is more robust and predictable if we add the `pyenv.cmd` in the front.  This guarantees that all (import) actions take place in this environment.

not-my-profile replied:
> but we don't know it

I am 99% sure.

Pyright has absolutely no dependencies on Python (that's the reason they wrote it in TypeScript).

Running it in a venv when it's completely unnecessary just doesn't make sense (it's even misleading).

return42 replied:
> Pyright has absolutely no dependencies on Python (that's the reason they wrote it in TypeScript).

I have a doubt .. TBH I don't know the details but I assume there are dependencies, at least the python version is a dependence.

not-my-profile replied:
>at least the python version is a dependence.

No it's not, see the [docs](https://github.com/microsoft/pyright/blob/main/docs/configuration.md):

>pythonVersion [string, optional]: Specifies the version of Python that will be used to execute the source code. The version should be specified as a string in the format "M.m" where M is the major version and m is the minor (e.g. "3.0" or "3.6"). If a version is provided, pyright will generate errors if the source code makes use of language features that are not supported in that version. It will also tailor its use of type stub files, which conditionalizes type definitions based on the version. If no version is specified, pyright will use the version of the current python interpreter, if one is present.

The question is how the current python interpreter is determined :thinking:

return42 replied:
> The version should be specified as a string in the format "M.m" 

Thanks for clarify!

> The question is how the current python interpreter is determined thinking

Thats why I thought it is better we integrate the CI test in the python tests:

https://github.com/searxng/searxng/blob/96a1f79c6dcaa08a231ad3a28b8c17b677cddc9c/.github/workflows/integration.yml#L16

not-my-profile replied:
Ah ok ... we were both correct :P 

You were right that we should execute it in the env and I was right that pyright doesn't need to be run in the env. But we still should because it executes whatever `python` is on the PATH to determine the Python version.

I'll rebase the PR.

return42 replied:
> Ah ok ... we were both correct :P

:-)  :-) ...  

>  If no version is specified, pyright will use the version of the current python interpreter, if one is present.

When we run the tests in the different python version, I think we do not need to define a version for pyright.

> I'll rebase the PR.

Thanks! .. I'am looking forward :-)

not-my-profile committed:
[mod] add test.pyright to test & ci.test targets

Since we currently have many type checking errors,
we for now only test with typeCheckingMode: off
which makes pyright only check files that contain a comment:

    # pyright: basic

to enable basic type checking, or

    # pyright: strict

to enable strict type checking.

not-my-profile committed:
[fix] ensure that test.pyright installs pyright

nodejs.ensure only sets up NVM if there isn't a system-wide
installation of Node that matches our NODE_MINIMUM_VERSION.

The ubuntu image in the CI comes with an up to date node version,
so pyright from .nvm_packages is never installed.

This commit fixes this by introducing a package.json file.

return42 commented on manage:
```diff
@@ -682,6 +683,18 @@ test.pylint() {
+    [ ! -d node_modules ] && npm install
+    # We run Pyright in the virtual environment because Pyright
+    # executes "python" to determine the Python version.
+    pyenv.cmd npx --no-install pyright -p pyrightconfig-ci.json
```
If we here use `pyrightconfig-ci.json` where do we use `pyrightconfig.json`? --> why not adding `"typeCheckingMode": "off"` to the `pyrightconfig.json` and remove `pyrightconfig-ci.json`?

not-my-profile replied:
The `pyrightconfig.json` is there because it's automatically detected by the Pyright extension for VS Code (and I assume also Pyright extensions for other editors). I did not set `"typeCheckingMode": "off"` in `pyrightconfig.json` because when viewing the files in your editor you want all errors to be reported (so that you can fix them).

return42 replied:
> because when viewing the files in your editor you want all errors to be reported 

Ah, OK .. make sense!

not-my-profile committed:
[mod] remove .nvm_packages, add eslint to package.json

return42 committed:
[mod] introduce node.env.devtools function

not-my-profile commented on utils/lib_nvm.sh:
```diff
@@ -117,7 +117,9 @@ nvm.install() {
     git checkout "${NVM_VERSION_TAG}" 2>&1 | prefix_stdout "  ${_Yellow}||${_creset} "
     popd &> /dev/null
-    cp "${REPO_ROOT}/.nvm_packages" "${NVM_DIR}/default-packages"
+    if [ -d "${REPO_ROOT}/.nvm_packages" ]; then
```
Isn't this supposed to be `-f`?

return42 replied:
Oops .. yes!

return42 commented on manage:
```diff
@@ -20,6 +20,8 @@ source "$(dirname "${BASH_SOURCE[0]}")/utils/lib_go.sh"
 # shellcheck source=utils/lib_redis.sh
 source "$(dirname "${BASH_SOURCE[0]}")/utils/lib_redis.sh"
 
+PATH="${REPO_ROOT}/node_modules/.bin;${PATH}"
```
Ugly typo in path separator from mine:  `;` --> `:`

not-my-profile replied:
Damn that's a nasty typo! Was that the cause of your NVM problems?

return42 replied:
yes

not-my-profile replied:
Well, I'm glad that it wasn't anything more complicated!
